### ISSUE TITLE
0.5.1版本会在后台（win10）卡住，无法切换出来

### ISSUE THREAD

**Issue by gansui** [rust-kotlin/ashell#119]
运行一段时间后，随机，点击ashell在任务栏上的图标，无法切换到前台，而且ctrl+tab切换任务也无法把当前窗口 切换为ashell

**Comment by gansui** [comment]
是否#118已经解决这个问题？0.5.1版本不行

**Comment by gansui** [comment]
我用AI分析并编译测试后，问题就解决了，改动很小,只改了一处 src/app/mod.rs:1285：
  经过对ashell代码和GPUI框架（Zed editor的UI框架）的Windows平台实现的深入分析，
     问题很可能出在以下几个方面：

     1. GPUI的activate()方法在任务栏点击时未被调用

     查看GPUI Windows平台的窗口激活流程（gpui_windows/src/window.rs）：

     fn activate(&self) {
         // 1. 恢复窗口位置
         this.set_window_placement();
         // 2. 如果最小化了则恢复
         if IsIconic(hwnd).as_bool() { ShowWindowAsync(hwnd, SW_RESTORE); }
         // 3. 设置焦点
         SetActiveWindow(hwnd); SetFocus(Some(hwnd));
         // 4. 关键hack：模拟VK_MENU按键绕过Windows前台窗口限制
         SendInput(&inputs, ...);  // VK_MENU keydown + keyup
         // 5. 最终调用SetForegroundWindow
         SetForegroundWindow(hwnd);

    }                                                                                                       47% used
                                                                                                             $0.00 spent
     这个activate()方法包含了一个关键的VK_MENU按键模拟hack来绕过Windows的前台窗口限制。但是：
                                                                                                             LSP
     - 这个方法只在ashell代码主动调用window.activate_window()时才会执行                                      • yaml-ls
     - 当前代码中，window.activate_window()只在两处被调用：
       - startup.rs:342：窗口创建时
       - mod.rs:1191：点击通知时
     - 当用户点击任务栏时，Windows发送WM_ACTIVATE消息，GPUI的handle_activate_msg只返回None让
       DefWindowProcW处理，但没有调用PlatformWindow::activate()，因此VK_MENU hack和SetForegroundWindow
       都不会被执行

     2. GPUI的WM_ACTIVATE处理有潜在问题

     GPUI的WM_ACTIVATE处理（events.rs:730-777）是异步的：

     fn handle_activate_msg(self: &Rc<Self>, wparam: WPARAM) -> Option<isize> {
         let activated = wparam.loword() > 0;
         // ...更新无障碍状态、重置光标和修饰键
         self.executor.spawn(async move {
             // 异步回调！
             func(activated);
         }).detach();
         None  // 交给DefWindowProcW
     }

    返回None意味着交给DefWindowProcW处理。在正常情况下，DefWindowProcW应该能处理好WM_ACTIVATE。但在
     Windows的前台窗口限制下，DefWindowProcW可能无法将窗口带到前台。

     3. Windows前台窗口锁定机制

     Windows有一个保护机制：进程不能随意调用SetForegroundWindow，除非满足特定条件（如最近收到过输入事件、
     是前台进程等）。GPUI的activate()方法通过模拟VK_MENU按键来绕过这个限制，但这个hack只在activate()
     被调用时才生效。

     4. ashell的事件循环中有冗余的激活检查

     src/app/mod.rs:1306-1351中的事件循环每16ms调用一次sync_window_activation：

     cx.spawn_in(window, async move |this, cx| {
         loop {
             cx.background_executor().timer(Duration::from_millis(16)).await;
             // ...
             let activation_changed = this.sync_window_activation(window);
             // ...


     这不会造成问题，但sync_window_activation只是读取window.is_window_active()来更新内部状态，
     不会主动去激活窗口。

     修复方案

     问题的修复方向应该是在on_window_activation_changed回调中，当检测到窗口变为激活状态时，主动调用
     window.activate_window()来确保窗口被带到前台。

     具体修改位置：src/app/mod.rs中的on_window_activation_changed方法（第1285行）。

     需要修改的代码：

     fn on_window_activation_changed(&mut self, window: &mut Window, cx: &mut Context<Self>) {
         self.sync_window_activation(window);
         cx.notify();
     }

     可能需要改为在窗口变为激活状态时调用window.activate_window()，
     但这需要验证不会导致无限循环或其他副作用。



只改了一处 src/app/mod.rs:1285：
     fn on_window_activation_changed(&mut self, window: &mut Window, cx: &mut Context<Self>) {
         let was_active = self.window_active;
         self.sync_window_activation(window);
         if !was_active && self.window_active {
             window.activate_window();  // 新增
         }
         cx.notify();
     }

     原理：当窗口从非激活变为激活状态时，主动调用 window.activate_window()。GPUI 的 activate()
     方法内部包含一个 VK_MENU 按键模拟 hack 和 SetForegroundWindow 调用，可以绕过 Windows
     的前台窗口锁定机制，确保窗口被带到前台。

**Comment by realmx** [comment]
> 我用AI分析并编译测试后，问题就解决了，改动很小,只改了一处 src/app/mod.rs:1285： 经过对ashell代码和GPUI框架（Zed editor的UI框架）的Windows平台实现的深入分析， 问题很可能出在以下几个方面：
> 
> ```
>  1. GPUI的activate()方法在任务栏点击时未被调用
> 
>  查看GPUI Windows平台的窗口激活流程（gpui_windows/src/window.rs）：
> 
>  fn activate(&self) {
>      // 1. 恢复窗口位置
>      this.set_window_placement();
>      // 2. 如果最小化了则恢复
>      if IsIconic(hwnd).as_bool() { ShowWindowAsync(hwnd, SW_RESTORE); }
>      // 3. 设置焦点
>      SetActiveWindow(hwnd); SetFocus(Some(hwnd));
>      // 4. 关键hack：模拟VK_MENU按键绕过Windows前台窗口限制
>      SendInput(&inputs, ...);  // VK_MENU keydown + keyup
>      // 5. 最终调用SetForegroundWindow
>      SetForegroundWindow(hwnd);
> 
> }                                                                                                       47% used
>                                                                                                          $0.00 spent
>  这个activate()方法包含了一个关键的VK_MENU按键模拟hack来绕过Windows的前台窗口限制。但是：
>                                                                                                          LSP
>  - 这个方法只在ashell代码主动调用window.activate_window()时才会执行                                      • yaml-ls
>  - 当前代码中，window.activate_window()只在两处被调用：
>    - startup.rs:342：窗口创建时
>    - mod.rs:1191：点击通知时
>  - 当用户点击任务栏时，Windows发送WM_ACTIVATE消息，GPUI的handle_activate_msg只返回None让
>    DefWindowProcW处理，但没有调用PlatformWindow::activate()，因此VK_MENU hack和SetForegroundWindow
>    都不会被执行
> 
>  2. GPUI的WM_ACTIVATE处理有潜在问题
> 
>  GPUI的WM_ACTIVATE处理（events.rs:730-777）是异步的：
> 
>  fn handle_activate_msg(self: &Rc<Self>, wparam: WPARAM) -> Option<isize> {
>      let activated = wparam.loword() > 0;
>      // ...更新无障碍状态、重置光标和修饰键
>      self.executor.spawn(async move {
>          // 异步回调！
>          func(activated);
>      }).detach();
>      None  // 交给DefWindowProcW
>  }
> 
> 返回None意味着交给DefWindowProcW处理。在正常情况下，DefWindowProcW应该能处理好WM_ACTIVATE。但在
>  Windows的前台窗口限制下，DefWindowProcW可能无法将窗口带到前台。
> 
>  3. Windows前台窗口锁定机制
> 
>  Windows有一个保护机制：进程不能随意调用SetForegroundWindow，除非满足特定条件（如最近收到过输入事件、
>  是前台进程等）。GPUI的activate()方法通过模拟VK_MENU按键来绕过这个限制，但这个hack只在activate()
>  被调用时才生效。
> 
>  4. ashell的事件循环中有冗余的激活检查
> 
>  src/app/mod.rs:1306-1351中的事件循环每16ms调用一次sync_window_activation：
> 
>  cx.spawn_in(window, async move |this, cx| {
>      loop {
>          cx.background_executor().timer(Duration::from_millis(16)).await;
>          // ...
>          let activation_changed = this.sync_window_activation(window);
>          // ...
> 
> 
>  这不会造成问题，但sync_window_activation只是读取window.is_window_active()来更新内部状态，
>  不会主动去激活窗口。
> 
>  修复方案
> 
>  问题的修复方向应该是在on_window_activation_changed回调中，当检测到窗口变为激活状态时，主动调用
>  window.activate_window()来确保窗口被带到前台。
> 
>  具体修改位置：src/app/mod.rs中的on_window_activation_changed方法（第1285行）。
> 
>  需要修改的代码：
> 
>  fn on_window_activation_changed(&mut self, window: &mut Window, cx: &mut Context<Self>) {
>      self.sync_window_activation(window);
>      cx.notify();
>  }
> 
>  可能需要改为在窗口变为激活状态时调用window.activate_window()，
>  但这需要验证不会导致无限循环或其他副作用。
> ```
> 
> 只改了一处 src/app/mod.rs:1285： fn on_window_activation_changed(&mut self, window: &mut Window, cx: &mut Context) { let was_active = self.window_active; self.sync_window_activation(window); if !was_active && self.window_active { window.activate_window(); // 新增 } cx.notify(); }
> 
> ```
>  原理：当窗口从非激活变为激活状态时，主动调用 window.activate_window()。GPUI 的 activate()
>  方法内部包含一个 VK_MENU 按键模拟 hack 和 SetForegroundWindow 调用，可以绕过 Windows
>  的前台窗口锁定机制，确保窗口被带到前台。
> ```

你可以提交PR等作者审核（PS：作者太懒了，以至于我想自己开库）

**Comment by gansui** [comment]
你要是想用这个版本可以到https://github.com/gansui/ashell的action中下载，我只编译了windows版本,下载的链接:
https://github.com/gansui/ashell/actions/runs/34007045906/artifacts/9981387397

**Comment by realmx** [comment]
> 你要是想用这个版本可以到[https://github.com/gansui/ashell的action中下载，我只编译了windows版本,下载的链接](https://github.com/gansui/ashell%E7%9A%84action%E4%B8%AD%E4%B8%8B%E8%BD%BD%EF%BC%8C%E6%88%91%E5%8F%AA%E7%BC%96%E8%AF%91%E4%BA%86windows%E7%89%88%E6%9C%AC,%E4%B8%8B%E8%BD%BD%E7%9A%84%E9%93%BE%E6%8E%A5): https://github.com/gansui/ashell/actions/runs/34007045906/artifacts/9981387397

哎。。不等了，我自己重构一个，给自己用

**Comment by rust-kotlin** [comment]
我的显示器在返厂维修，主机不太方便使用。

@realmx 我已经将你添加为协作者了，你可以自行管理pr与提交。
